iterator_test.go

```diff
@@ -2737,3 +2737,176 @@ func BenchmarkSeekPrefixTombstones(b *testing.B) {
+		return key[:prefixLen+n]
+	}
+
+	type Queue struct {
+		start int
+		end   int // exclusive
+	}
+	var queues = make([]*Queue, queueCount)
+	for i := 0; i < queueCount; i++ {
+		queues[i] = &Queue{}
+	}
+
+	o := (&Options{
+		DisableWAL:         true,
+		FS:                 vfs.NewMem(),
+		Comparer:           testkeys.Comparer,
+		FormatMajorVersion: FormatNewest,
+	}).EnsureDefaults()
+
+	d, err := Open("", o)
+	require.NoError(b, err)
+
+	processQueueOnce := func(batch *Batch, batchSize int) *Batch {
+		for {
+			// Randomly pick a queue to process.
+			q := rng.Intn(queueCount)
+			queue := queues[q]
+
+			isDelete := rng.Float32() < deleteRatio
+
+			if isDelete {
+				// Only process the queue if it's not empty. Otherwise, retry
+				// with a different queue.
+				if queue.start < queue.end {
+					require.NoError(b, batch.Delete(getKey(q, queue.start), nil))
+					queue.start++
+					break
+				}
+			} else {
+				// Append to the queue.
+				require.NoError(b, batch.Set(getKey(q, queue.end), val, nil))
+				queue.end++
+				break
+			}
+		}
+
+		if batch.Len() >= batchSize {
+			require.NoError(b, batch.Commit(NoSync))
+			d.AsyncFlush()
+			return d.NewBatch()
+		}
+		return batch
+	}
+
+	// First, process queues initialOps times.
+	batch := d.NewBatch()
+	for i := 0; i < initOps; i++ {
+		// Use a large batch size to speed up initialization.
+		batch = processQueueOnce(batch, 10<<24 /* 167 MiB */)
+	}
+	require.NoError(b, batch.Commit(NoSync))
+
+	b.Run(fmt.Sprintf("initial_ops=%d,deleteRatio=%.2f,valueSize=%d", initOps, deleteRatio, valueSize), func(b *testing.B) {
+		waitForCompactions := func() {
+			d.mu.Lock()
+			// NB: Wait for table stats because some compaction types rely
+			// on table stats to be collected.
+			d.waitTableStats()
+			for d.mu.compact.compactingCount > 0 {
+				d.mu.compact.cond.Wait()
+				d.waitTableStats()
+			}
+			d.mu.Unlock()
+		}
+
+		waitForCompactions()
+
+		// Log the number of tombstones and live keys in each level after
+		// background compactions are complete.
+		b.Log("LSM after compactions:")
+		firstIter, _ := d.NewIter(nil)
+		firstIter.First()
+		lastIter, _ := d.NewIter(nil)
+		lastIter.Last()
+		stats, _ := d.ScanStatistics(context.Background(), firstIter.Key(), lastIter.Key(), ScanStatisticsOptions{})
+		require.NoError(b, firstIter.Close())
+		require.NoError(b, lastIter.Close())
+		metrics := d.Metrics()
+		for i := 0; i < numLevels; i++ {
+			numTombstones := stats.Levels[i].KindsCount[base.InternalKeyKindDelete]
+			numSets := stats.Levels[i].KindsCount[base.InternalKeyKindSet]
+			numTables := metrics.Levels[i].NumFiles
+			if numSets > 0 {
+				b.Logf("L%d: %d tombstones, %d sets, %d sstables\n", i, numTombstones, numSets, numTables)
+			}
+		}
+
+		b.Run("queue_throughput", func(b *testing.B) {
+			batch := d.NewBatch()
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				batch = processQueueOnce(batch, 10<<10 /* 10 KiB */)
```

Comment:
The last batch isn't getting committed, right? You could consider moving the batch commit outside of `processQueueOnce` just to make it more readable.

Reply:
Yeah, the last batch wasn't being used - I've updated to move the commit logic outside